### DISCUSSION TITLE
Get logs after calling contract.write with viem

### DISCUSSION THREAD

**Discussion by pepoospina** [NomicFoundation/hardhat#4573] [Troubleshooting]
I use 
```ts
cont client = await viem.getPublicClient();
const contract = await viem.deployContract()
const hash = await contract.write.method([])` 
const rec = await client.waitForTransactionReceipt({ hash });
``` 

But the logs are not parsed as events. Is there a way to parse them? these are my contracts tests. I don't have access to abi (or I don't know how to access the abi, and hoped I should not need to do it since hardhat knows all about my contracts already)

**Comment by pepoospina** [comment]
Maybe there is an automatic way to do this, but this way around should do it too

```ts
import { artifacts } from 'hardhat';
import { decodeEventLog } from 'viem';

type EventType = Awaited<ReturnType<typeof decodeEventLog>>;

const getContractEventsFromReceipt = async (contractName: string, rec: TransactionReceipt): Promise<EventType[]> => {
  const factoryArtifact = await artifacts.readArtifact(contractName);
  if (!rec.logs) return [];
  const events = rec.logs
    .map((log) => {
      try {
        const event = decodeEventLog({
          abi: factoryArtifact.abi,
          data: log.data,
          topics: log.topics,
          strict: false,
        });
        return event;
      } catch (e) {
        return undefined;
      }
    })
    .filter((e) => e !== undefined);

  return events as unknown as EventType[];
};
```

**Comment by fvictorio** [comment]
Can't you use `contract.getEvents.MyEvent()`?

**Reply by pepoospina** [reply]
It would be another option, but would still need some additional manipulation to scope to one single block, then from there filter for a specific transaction. 

I believe the "logs of a tx" is a readily available data attached to the tx data itself. Filtering Events based on block number and then based on tx hash, would be much slower, I'd guess.

This seems more like a Viem limitation. It seems a `.writeAndWaitForTransactionReceipt()` method on a contract instance would be needed so that the contract context under which the receipt is read is known.
